synapse/handlers/oauth2_handler.py

```diff
@@ -0,0 +1,154 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+
+from six.moves import urllib
+
+from twisted.web.client import PartialDownloadError
+
+from synapse.api.errors import Codes, LoginError
+from synapse.http.servlet import parse_string
+from synapse.rest.client.v1.login import SSOAuthHandler
+from synapse.util.caches.expiringcache import ExpiringCache
+from synapse.util.stringutils import random_string
+
+logger = logging.getLogger(__name__)
+
+
+class OAuth2Handler:
+    def __init__(self, hs):
+        # config
+        self.public_baseurl = hs.config.public_baseurl.encode("ascii")
+        self.oauth2_server_authorization_url = hs.config.oauth2_server_authorization_url.encode(
+            "ascii"
+        )
+        self.oauth2_server_token_url = hs.config.oauth2_server_token_url
+        self.oauth2_server_userinfo_url = hs.config.oauth2_server_userinfo_url
+        self.oauth2_client_id = hs.config.oauth2_client_id
+        self.oauth2_client_secret = hs.config.oauth2_client_secret
+        self.oauth2_scope = "openid"
+        self.oauth2_response_type = "code"
+        self.oauth2_response_mode = "query"
+
+        # state
+        self.nonces = ExpiringCache(
+            cache_name="oauth_nonces",
+            clock=hs.get_clock(),
+            expiry_ms=5 * 60 * 1000,  # 5 minutes
+            reset_expiry_on_get=False,
+        )
+
+        # tools
+        self._sso_auth_handler = SSOAuthHandler(hs)
+        self._http_client = hs.get_proxied_http_client()
+
+    def handle_redirect_request(self, client_redirect_url):
+        """Handle an incoming request to /login/sso/redirect
+
+        Args:
+            client_redirect_url (bytes): the URL that we should redirect the
+                client to when everything is done
+
+        Returns:
+            bytes: URL to redirect to
+        """
+
+        oauth2_nonce = random_string(12)
+        self.nonces[oauth2_nonce] = {"redirectUrl": client_redirect_url}
```

Comment:
the `nonce` and the `state` fields should be two different things and should be checked somehow when the clients return (usually with a cookie). Also, I'm not sure if storing that in memory is a good idea, as it means this servlet can't be scaled, and restarting it will make all the running auth flows fail.